frame/migrations/src/lib.rs

```diff
@@ -0,0 +1,488 @@
+	pub enum Event<T: Config> {
+		/// Runtime upgrade started with `migrations` in the queue.
+		///
+		/// This can be used to design a progress indicator in combination with counting the
+		/// `MigrationCompleted` and `MigrationSkippedHistoric` events.
+		UpgradeStarted { migrations: u32 },
+		/// Runtime upgrade completed with `migrations`.
+		UpgradeCompleted,
+		/// Runtime upgrade failed.
+		///
+		/// This is very bad and will require governance intervention.
+		UpgradeFailed,
+		/// Migration `index` was skipped, since it already executed in the past.
+		MigrationSkippedHistoric { index: u32 },
+		/// Migration `index` made progress.
+		MigrationAdvanced { index: u32, blocks: T::BlockNumber },
+		/// Migration `index` completed.
+		MigrationCompleted { index: u32, blocks: T::BlockNumber },
+		/// Migration `index` failed.
+		///
+		/// This implies that the whole upgrade failed and governance intervention is required.
+		MigrationFailed { index: u32, blocks: T::BlockNumber },
+		/// The list of historical migrations has been cleared.
+		HistoricCleared {
+			/// Should be passed to `clear_historic` in a successive call.
+			next_cursor: Option<Vec<u8>>,
+		},
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			Self::start_mbm()
+		}
+
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			Self::progress_mbm(n)
+		}
+	}
+
+	#[pallet::call(weight = T::WeightInfo)]
+	impl<T: Config> Pallet<T> {
+		/// Allows root to set the cursor to any value.
+		///
+		/// Should normally not be needed and is only in place as emergency measure.
+		#[pallet::call_index(0)]
+		pub fn force_set_cursor(
+			origin: OriginFor<T>,
+			cursor: Option<MigrationCursor<T::Cursor, T::BlockNumber>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			Cursor::<T>::set(cursor);
+
+			Ok(())
+		}
+
+		/// Clears the `Historic` set.
+		///
+		/// `map_cursor` must be set to the last value that was returned by the
+		/// `HistoricCleared` event. The first time `None` can be used. `limit` must be chosen in a
+		/// way that will result in a sensible weight.
+		#[pallet::call_index(1)]
+		#[pallet::weight({0})] // FAIL-CI
+		pub fn clear_historic(
+			origin: OriginFor<T>,
+			limit: Option<u32>,
+			map_cursor: Option<Vec<u8>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let next = Historic::<T>::clear(limit.unwrap_or_default(), map_cursor.as_deref());
+			Self::deposit_event(Event::HistoricCleared { next_cursor: next.maybe_cursor });
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Starts the process of multi block migrations.
+	///
+	/// Should only be called once all migrations completed.
+	fn start_mbm() -> Weight {
+		use FailedUpgradeHandling::*;
+
+		if let Some(cursor) = Cursor::<T>::get() {
+			log::error!(target: LOG, "Ongoing migrations interrupted - chain stuck");
+			Self::deposit_event(Event::UpgradeFailed);
```

Comment:
Okay I am getting this more now: You are assuming that The cursor type is fixed, known, and all migrations are done before the next batch is possibly onboarded in the next runtime upgrade.